server.py

```
@@ -1,58 +1,78 @@
+    def __init__(self, boards: 'dict that maps room id to CanvasBoard'):
+        self.boards = boards
 
+    def update_board(self, diffs: list, room_id: str):
+        self.boards[room_id].update_board(diffs)
 
 # instantiate server class for board state
 imagedata = {
     'width': WIDTH,
     'height': HEIGHT,
     'data': [0 for i in range(4 * WIDTH * HEIGHT)]
 }
-server = Server(imagedata)
+server = Server({})
 
 @app.route('/')
 def index():
     return app.send_static_file('index.html')
 
+@app.route('/create')
+def create():
+    room = None
+    while room in server.boards or room is None:
+        room = ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(ID_LENGTH))
+        print(room)
+    response = jsonify(room_id=room)
+    # just return response with room id
+    # don't worry about frontend
+    return response
+
 @socketio.on('connect', namespace='/canvas')
 def connect_canvas():
-    # broadcast board upon initial connect at /canvas endpoint
-    # turn board into JSON
-    board = CanvasBoardEncoder().encode(server.board)
-    emit('initialize-board', board)
+    print("connected to websocket")
 
 @socketio.on('disconnect')
 def on_disconnect():
     print("disconnected from websocket")
 
 @socketio.on('send-stroke', namespace="/canvas")
-def handle_send_stroke(stroke):
-    server.update_board(stroke['diffs'])
-    # broadcast the new change
-    emit('broadcast-stroke', stroke, broadcast=True)
+def handle_send_stroke(stroke, room_data):
+    room_id = room_data['room_id']
+    if room_id in server.boards:
+        server.update_board(stroke['diffs'], room_id)
+        emit('broadcast-stroke', stroke, room=room_id)
+    else:
+        print("Error: no room found")
+
+@socketio.on('join', namespace='/canvas')
+def on_join(room_data):
+    room_id = room_data['room_id']
+    # instantiate initial CanvasBoard
+    if room_id not in server.boards:
+        server.boards[room_id] = CanvasBoard(imagedata)

```

We should fail loudly if the room the client is trying to join does not exist.

Refer to https://flask.palletsprojects.com/en/1.1.x/quickstart/#redirects-and-errors

---

Abort() function called if room_id not found